### PR TITLE
[SPARK-49306][SQL] Create new SQL functions 'zeroifnull' and 'nullifzero' 

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR creates new SQL functions `zeroifnull` and `nullifzero`.

These are equivalent to `nvl(input, 0)` and `if(input = 0, null, input)`, respectively. The analyzer performs these replacements inline in this PR to take advantage of existing query optimizations for `nvl` and `if` expressions.

For example:

```
> select zeroifnull(null);
 0
> select zeroifnull(2);
 2

> select nullifzero(0);
 null
> select nullifzero(2);
 2
```

### Why are the changes needed?

We are adding these functions for compatibility with other systems which also have these SQL functions.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds unit test coverage.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @HyukjinKwon @MaxGekk can one of you guys maybe help with this one :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 1]
```diff
@@ -384,6 +385,7 @@ object FunctionRegistry {
     expression[Rand]("random", true, Some("3.0.0")),
     expression[Randn]("randn"),
     expression[Stack]("stack"),
+    expression[ZeroIfNull]("zeroifnull"),
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:388]: should probably add them into functions.scala and functions.py. could be done in a separate PR tho.

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/MiscFunctionsSuite.scala [hunk 2]
```diff
@@ -285,6 +285,48 @@ class MiscFunctionsSuite extends QueryTest with SharedSparkSession {
     assert(df.selectExpr("random(1)").collect() != null)
     assert(df.select(random(lit(1))).collect() != null)
   }
+
+  test("SPARK-49306 nullifzero and zeroifnull functions") {
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/MiscFunctionsSuite.scala:289]: This test suite is for misc functions mostly. Could you move expression tests to `ConditionalExpressionSuite` and end-to-end tests to `conditional-functions.sql`, please.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 1 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:388]: Thanks, I will add DataFrame support in a separate PR :)

File: sql/core/src/test/scala/org/apache/spark/sql/MiscFunctionsSuite.scala (hunk 2 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/MiscFunctionsSuite.scala]: Sure, it turns out all these were end to end tests so I just moved them all to `conditional-functions.sql`.

**Review by dtenedor — COMMENTED** [review]
Thanks @MaxGekk @HyukjinKwon for your reviews, please take another look.

**Review by allisonwang-db — APPROVED** [review]
LGTM!

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @dtenedor and @HyukjinKwon @allisonwang-db for review.

**Pull request closed without merging**